crates/compiler/mono/src/borrow.rs

```diff
@@ -1,119 +1,401 @@
+                }
+
+                std::mem::swap(&mut proc_join_points, &mut join_points[index]);
+
+                let mut state = State {
+                    args: proc.args,
+                    borrow_signature: *borrow_signatures.procs.get(&key).unwrap(),
+                    join_point_stack,
+                    join_points: proc_join_points,
+                    modified: false,
+                };
+
+                state.inspect_stmt(interner, &mut borrow_signatures, &proc.body);
+
+                // did any proc signature get modified?
+                //
+                // NOTE: this does not directly include updates to join point signatures. The
+                // assumption is that a relevant change in join point signature is immediately
+                // (i.e. no fixpoint is required) reflected in the proc signature.
+                assert_eq!(
+                    state.modified,
+                    borrow_signatures
+                        .procs
+                        .insert(key, state.borrow_signature)
+                        .unwrap()
+                        != state.borrow_signature
+                );
+                modified |= state.modified;
+
+                proc_join_points = state.join_points;
+
+                std::mem::swap(&mut proc_join_points, &mut join_points[index]);
+
+                join_point_stack = state.join_point_stack;
+                join_point_stack.clear();
+            }
+
+            if !modified {
+                break;
+            }
+        }
+    }
+
+    borrow_signatures
+}
+
 #[allow(unused)]
-pub(crate) fn infer_borrow_signature<'a>(
+fn infer_borrow_signature<'a>(
     arena: &'a Bump,
     interner: &impl LayoutInterner<'a>,
-    proc: &'a Proc<'a>,
-) -> &'a [Ownership] {
-    let mut state = State::new(arena, interner, proc);
-    state.inspect_stmt(&proc.body);
+    borrow_signatures: &'a mut BorrowSignatures<'a>,
+    proc: &Proc<'a>,
+) -> BorrowSignature {
+    let mut state = State::new(arena, interner, borrow_signatures, proc);
+    state.inspect_stmt(interner, borrow_signatures, &proc.body);
     state.borrow_signature
 }
 
-struct State<'a> {
+struct State<'state, 'arena> {
     /// Argument symbols with a layout of `List *` or `Str`, i.e. the layouts
     /// for which borrow inference might decide to pass as borrowed
-    args: &'a [(InLayout<'a>, Symbol)],
-    borrow_signature: &'a mut [Ownership],
+    args: &'state [(InLayout<'arena>, Symbol)],
+    borrow_signature: BorrowSignature,
+    join_point_stack: Vec<'arena, (JoinPointId, &'state [Param<'arena>])>,
+    join_points: MutMap<JoinPointId, BorrowSignature>,
+    modified: bool,
 }
 
 fn layout_to_ownership<'a>(
     in_layout: InLayout<'a>,
     interner: &impl LayoutInterner<'a>,
 ) -> Ownership {
     match interner.get_repr(in_layout) {
-        LayoutRepr::Builtin(Builtin::Str | Builtin::List(_)) => Ownership::Borrowed,
+        LayoutRepr::Builtin(Builtin::Str) => Ownership::Borrowed,
```

Comment:
Do we need to an something recursive so this also works on a `Dict` and `Set`? Like borrowing the `Dict`/`Set` as a while. I'm not worrying about trying to do borrow inference for subfields in a struct separately.

Though I guess that still would require borrow inference through loading from a record, which might require more complex wiring.

Reply:
I don't really follow. If a `Str` or `List` layout is passed directly to a function (or in a lambda set) then it may be borrowed now, anything more complex is not considered.

Reply:
Ok. Yeah, so doesn't support `Dict` but once we unwrap the `Dict` and start passing around its inner state list, it should work.

So this PR has no borrow inference for the `Dict` passed into [find](https://github.com/roc-lang/roc/blob/a44a6889bd6a5cbd4835eb399d6d2342d7e00947/crates/compiler/builtins/roc/Dict.roc#L735-L746), but borrow inference for [these inner functions](https://github.com/roc-lang/roc/blob/a44a6889bd6a5cbd4835eb399d6d2342d7e00947/crates/compiler/builtins/roc/Dict.roc#L750-L788). Which I think should still get all refcounting out of the inner loop which will still greatly enhance performance.